### ISSUE TITLE
[Request] Pixai, CL, RedRocket tagger and character ip/copyright mapping file

### ISSUE THREAD

**Issue by DraconicDragon** [hydrui/hydrui#14]
I'd like to suggest the following tagging models and also point to the possibility of adding the Pixai tagger's character-ip/copyright mapping file as an independent mapping file  
For all the models more information can be found on the linked HuggingFace repositories

### Pixai Tagger and mapping file:
The Pixai tagger v0.9 is a finetune on top of the WD eva02 large tagger v3 model from SmilingWolf with updated knowledge and slightly better accuracy (at least in my experience), _however it should be noted that it doesn't support outputting rating tags_.

The mapping file could be useful in general regardless of model though I haven't done extensive testing for this if it's useful or not

Links (im unsure which format is needed): 
- ONNX [deepghs/pixai-tagger-v0.9-onnx](https://huggingface.co/deepghs/pixai-tagger-v0.9-onnx) | [Demo](https://huggingface.co/spaces/deepghs/pixai-tagger-v0.9-demo)
- Pytorch [pixai-labs/pixai-tagger-v0.9](https://huggingface.co/pixai-labs/pixai-tagger-v0.9) [Gated Repo] | [Demo](https://huggingface.co/spaces/pixai-labs/pixai-tagger-demo)
- **character ip mapping file**: [char_ip_map.json](https://huggingface.co/pixai-labs/pixai-tagger-v0.9/blob/main/char_ip_map.json) [Gated Repo, ~20mb JSON]
	- [selected_tags.csv](https://huggingface.co/deepghs/pixai-tagger-v0.9-onnx/blob/main/selected_tags.csv) [ungated, ~580KB CSV - the IPs are in a separate column]
	- I assume the CSV file can be further reduced in size for independent use by removing rows with an empty array in the IPs column

### CL Tagger:

It is like the Pixai tagger - a finetune on top of the WD eva02 model, however it outputs more categories than the former 2:

    General: 22,367 tags (general attributes and features)
    Character: 16,210 tags (character names)
    Copyright: 3,329 tags (series names)
    Meta: 224 tags (metadata)
    Model: 25 tags (model information)
    Rating: 4 tags (content rating)
    Quality: 4 tags (quality assessment)

In my experience it's not worth it over the PixAI tagger if accuracy is a priority because it seems to be on par or worse in general tag accuracy over eva02 (while considering that on the HF space demo, the default thresholds are increased) and meta, model and quality information probably aren't that useful (besides quality maybe, however I think that a specialized aesthetic classifier model would fit this task better).  
Even with the downsides I found it worthwhile to mention it

Link: ONNX [cella110n/cl_tagger](https://huggingface.co/cella110n/cl_tagger)
Demo: https://huggingface.co/spaces/cella110n/cl_tagger

### RedRocket Joint Tagger Project (e621 tags):

Unfortunately I can't comment on if it's better than Z3D-E621-Convnext and otherwise I'm not aware of other e621 tag tagging models so this has just been my go-to for tagging any relevant images

Links:
- ONNX [silveroxides/JTP_PILOT2-onnx](https://huggingface.co/silveroxides/JTP_PILOT2-onnx)
- Pytorch https://huggingface.co/deepghs/pixai-tagger-v0.9-onnx
Demo: https://huggingface.co/spaces/RedRocket/JointTaggerProject-Inference-Beta

**Comment by DraconicDragon** [comment]
Hi, 
first of all, sorry for late reply, I've been following the releases of this project and thank you for implementing pixai/cl tagger.  
Recently I've started experimenting with onnx conversion/export because of the release of a small artist style classifier and I also gave a try at converting JTP myself and uploaded the model files here: https://huggingface.co/DraconicDragon/JTP-PILOT2-onnx-ort    

I also have had AI write a quick and dirty [hugginface space demo](https://huggingface.co/spaces/DraconicDragon/JTP-PILOT2-onnx-ort-Demo) for this converted model to confirm its working where silveroxide's version can also be used for comparison and it seems like silveroxide's version gives seemingly random ids that don't match the csv ids - which of course may be due to the implementation so I plan to do some more tests on my own locally later on - but I thought maybe sharing and having you try my onnx/ort export might be a good idea.  

Though I should note, I believe I did something wrong during the pytorch -> onnx export because I did not specify dynamic_axes/dynamic_shapes (still unsure which one is the better/working option) so the model is likely locked to batch size 1

There's also some possibly interesting info/stuff I picked up but it's unlikely to be useful: 
- Just a few hours ago I found a swinv2 tune for e621 tags [here](https://huggingface.co/animetimm/swinv2_base_window8_256.e621v1-full) but it's a gated repo and I haven't requested access
- I recently learned about OpenVINO which seems to give a very nice boost for inference speed especially on (any) CPU and intel (i)GPUs with onnx models, but unfortunately it seems incompatible with onnxruntime for web/ort
- The [WD-swinv2 tagger v3](huggingface.co/SmilingWolf/wd-swinv2-tagger-v3) seems to be the most efficient model in terms of accuracy, speed and size; it achieves pretty much same results/accuracy as vit-large and outperforms vit, the currently recommended model, at their respective thresholds, however it is ~100mb larger than vit; speaking mostly from personal experience but the F1 scores on the readmes seem to back that up besides vit-large having a bit higher score than swinv2, though still lower than eva02 and pixai 0.9

**Comment by jchv** [comment]
Oh, awesome. I'll try to take a look at this soon. Regarding the batch size thing: right now Hydrui only ever uses a batch size of 1. To be honest, I wasn't entirely sure what batch size was intended to be used for, and not all models supported it, so I just ignored it for the time being. I'm still a pretty big n00b when it comes to machine learning technology, hence why I wasn't really able to debug why onnxruntime-web was crashing on the silveroxide version. I'm still definitely interested in adding new models to Hydrui, especially if they can push the frontier of efficiency and performance even further. (If not adding more models to the default list, then at least certainly adding more models to the models pseudorelease. But I have plenty of bandwidth to spare for serving more models off of models.hydrui.dev, so we can probably add more built-in models to select from.)

**Comment by jchv** [comment]
BTW, I did wind up trying your conversion of JTP PILOT2, but unfortunately it still seems to crash onnxruntime-web, so I guess whatever is causing that is not actually tied to exactly how the conversion is done, but is something deeper. I will probably need to find out how to build onnxruntime-web with assertions on at some point to figure it out.

I did wind up packaging WD-swinv2 tagger v3 and adding it to the models pseudo-release, which works very well.
